src/java/com/puppetlabs/ssl_utils/ExtensionsUtils.java

```diff
@@ -740,61 +741,92 @@ private static SubjectKeyIdentifier publicKeyToSubjectKeyIdentifier(PublicKey pu
+            if (pubKey == null && serialNumber == null) {
                 throw new IllegalArgumentException(
-                        "'serial_number' not provided for auth key identifier" +
-                        "but was expected since 'issuer' was provided");
+                        "Neither 'public_key', 'serial_number', or 'cert' provided for " +
+                        "auth key identifier.  At least one of these must be " +
+                        "provided.");
             }
 
-            GeneralNames issuerAsGeneralNames =
-                    new GeneralNames(new GeneralName(new X500Name(issuer)));
-            if (authorityKeyId != null) {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        authorityKeyId.getKeyIdentifier(),
-                        issuerAsGeneralNames,
-                        serialNumber);
+            String issuer = (String) authKeyIdMap.get ("issuer_dn");
+            if (issuer == null) {
+                if (serialNumber != null) {
+                    throw new IllegalArgumentException(
+                            "'issuer' not provided for auth key identifier " +
+                            "but was expected since 'serial_number' was provided");
+                }
             }
             else {
-                authorityKeyId = new AuthorityKeyIdentifier(
-                        issuerAsGeneralNames,
-                        serialNumber);
+                if (serialNumber == null) {
+                    throw new IllegalArgumentException(
+                            "'serial_number' not provided for auth key identifier" +
+                            "but was expected since 'issuer' was provided");
+                }
+
+                GeneralNames issuerAsGeneralNames =
+                        new GeneralNames(new GeneralName(new X500Name(issuer)));
+                if (authorityKeyId != null) {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            authorityKeyId.getKeyIdentifier(),
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
+                else {
+                    authorityKeyId = new AuthorityKeyIdentifier(
+                            issuerAsGeneralNames,
+                            serialNumber);
+                }
             }
         }
+        else {
+            DigestCalculator digCalc =
+                new JcaDigestCalculatorProviderBuilder().build().get(
+                        new AlgorithmIdentifier(OIWObjectIdentifiers.idSHA1));
+
+            X509ExtensionUtils utils = new X509ExtensionUtils(digCalc);
+            authorityKeyId = utils.createAuthorityKeyIdentifier(new X509CertificateHolder(cert.getEncoded()));
+
+            // byte[] extBytes = cert.getExtensionValue(Extension.subjectKeyIdentifier.toString());
+            // AuthorityKeyIdentifier authorityKeyId1 = new AuthorityKeyIdentifier(extBytes);
```

Comment:
I think you'd need to call `getBytes` on the string first?
You might want to see what they do in the section "And for decoding SKI:" in the stackoverflow:
https://stackoverflow.com/questions/31175683/how-to-know-which-x509-certificate-signed-another-certificate-java?noredirect=1&lq=1

Sorry this review took so long, I'm still playing with this locally.

Reply:
I looked at that code from that stackoverflow and I think I tried using octets from the `JcaX509CertificateHolder` and it matched up type-wise to generate a valid `AuthorityKeyIdentifier` but winds up adding an incorrect auth key id and puppet runs can't verify with SSL. I can try using `DEROctetString` to see if that does anything differently.